libkbfs/bare_root_metadata.go

```diff
@@ -0,0 +1,481 @@
+}
+
+// IsWriterMetadataCopiedSet returns true if the bit is set indicating
+// the writer metadata was copied.
+func (md *BareRootMetadata) IsWriterMetadataCopiedSet() bool {
+	return md.Flags&MetadataFlagWriterMetadataCopied != 0
+}
+
+// IsFinal returns true if this is the last metadata block for a given
+// folder.  This is only expected to be set for folder resets.
+func (md *BareRootMetadata) IsFinal() bool {
+	return md.Flags&MetadataFlagFinal != 0
+}
+
+// IsWriter returns whether or not the user+device is an authorized writer.
+func (md *BareRootMetadata) IsWriter(
+	user keybase1.UID, deviceKID keybase1.KID) bool {
+	if md.ID.IsPublic() {
+		for _, w := range md.Writers {
+			if w == user {
+				return true
+			}
+		}
+		return false
+	}
+	return md.WKeys.IsWriter(user, deviceKID)
+}
+
+// IsReader returns whether or not the user+device is an authorized reader.
+func (md *BareRootMetadata) IsReader(
+	user keybase1.UID, deviceKID keybase1.KID) bool {
+	if md.ID.IsPublic() {
+		return true
+	}
+	return md.RKeys.IsReader(user, deviceKID)
+}
+
+// updateNewRootMetadata initializes the given freshly-created
+// RootMetadata object with the given TlfID and TlfHandle. Note that
+// if the given ID/handle are private, rekeying must be done
+// separately.
+func updateNewRootMetadata(
+	rmd *BareRootMetadata, id TlfID, h BareTlfHandle) error {
+	if id.IsPublic() != h.IsPublic() {
+		return errors.New("TlfID and TlfHandle disagree on public status")
+	}
+
+	var writers []keybase1.UID
+	var wKeys TLFWriterKeyGenerations
+	var rKeys TLFReaderKeyGenerations
+	if id.IsPublic() {
+		writers = make([]keybase1.UID, len(h.Writers))
+		copy(writers, h.Writers)
+	} else {
+		wKeys = make(TLFWriterKeyGenerations, 0, 1)
+		rKeys = make(TLFReaderKeyGenerations, 0, 1)
+	}
+	rmd.WriterMetadata = WriterMetadata{
+		Writers: writers,
+		WKeys:   wKeys,
+		ID:      id,
+	}
+	if len(h.UnresolvedWriters) > 0 {
+		rmd.Extra.UnresolvedWriters = make([]keybase1.SocialAssertion, len(h.UnresolvedWriters))
+		copy(rmd.Extra.UnresolvedWriters, h.UnresolvedWriters)
+	}
+
+	rmd.Revision = MetadataRevisionInitial
+	rmd.RKeys = rKeys
+	if len(h.UnresolvedReaders) > 0 {
+		rmd.UnresolvedReaders = make([]keybase1.SocialAssertion, len(h.UnresolvedReaders))
+		copy(rmd.UnresolvedReaders, h.UnresolvedReaders)
+	}
+	return nil
+}
+
+func (md *BareRootMetadata) deepCopy(codec Codec) (*BareRootMetadata, error) {
+	var newMd BareRootMetadata
+	if err := md.deepCopyInPlace(codec, &newMd); err != nil {
+		return nil, err
+	}
+	return &newMd, nil
+}
+
+func (md *BareRootMetadata) deepCopyInPlace(codec Codec, newMd *BareRootMetadata) error {
+	if err := CodecUpdate(codec, newMd, md); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CheckValidSuccessor makes sure the given RootMetadata is a valid
+// successor to the current one, and returns an error otherwise.
+func (md *BareRootMetadata) CheckValidSuccessor(
+	currID MdID, nextMd *BareRootMetadata) error {
```

Comment:
Done, although there's one case where we still need to use the `BareRootMetadata` version (added comment for it, though).